README


This archive contains the distribution MooseX-Types-DateTime-MoreCoercions,
version 0.14:

  Extensions to L<MooseX::Types::DateTime>

This software is copyright (c) 2008 by John Napiorkowski.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



Changes
Revision history for MooseX-Types-DateTime-MoreCoercions

0.14      2014-08-06 06:04:21Z
        - re-release to remove README.pod from shipped dist

0.13      2014-07-30 14:23:01Z
        - fixed t/00-report-prereqs.t when CPAN::Meta is not installed

0.12      2014-07-21 14:51:38Z
        - repository has moved to the GitHub Moose organization
        - converted distribution packaging to Dist::Zilla, fixing mojibaked
          META.yml (RT#90138)
        - switch from using Test::Exception to Test::Fatal

0.11    2012-12-11 12:33:01Z
        - Require perl 5.8.3, as Moose does

0.10    2012-07-30 10:18:54Z
        - Remove unneeded check for DateTime::Format::DateManip from test
        - Rebuild with Module::Install 1.06, to avoid depending on dev
          release of EUMM (RT#77157)

0.09    2012-03-06 23:58:14Z
        - Use the introspection API to get MX:T:DateTime's coercions
        - fix a spelling mistake (RT#73343)

0.08    2011-12-12 13:18:23Z
        - Require fixed MooseX::Types::DateTime version (RT#73192)
        - Enforce version requirements at runtime (RT#73191)
        - Add missing dependencies (RT#73190)

0.07    2011-12-06 20:48:22Z
        - Switch back to using MooseX::Types::DateTime
        - Rename to more descriptive name

(imported from MooseX-Types-DateTimeX)


Build.PL
# This Build.PL for MooseX-Types-DateTime-MoreCoercions was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.005
if (eval 'use Module::Build::Tiny 0.037; 1')
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";

    # This section for MooseX-Types-DateTime-MoreCoercions was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.006.
    use 5.006;
    # use Module::Build::Tiny 0.037;
    Build_PL();
}
else
{
    $ENV{PERL_MB_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.

EOW
    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "build_requires" => {},
      "configure_requires" => {
        "ExtUtils::MakeMaker" => 0,
        "Module::Build::Tiny" => "0.037"
      },
      "dist_abstract" => "Extensions to L<MooseX::Types::DateTime>",
      "dist_author" => [
        "John Napiorkowski <jjn1056\@yahoo.com>",
        "Dagfinn Ilmari Manns\x{e5}ker <ilmari\@ilmari.org>"
      ],
      "dist_name" => "MooseX-Types-DateTime-MoreCoercions",
      "dist_version" => "0.14",
      "license" => "perl",
      "module_name" => "MooseX::Types::DateTime::MoreCoercions",
      "recommends" => {},
      "recursive_test_files" => 1,
      "requires" => {
        "DateTime" => "0.4302",
        "DateTime::Duration" => "0.4302",
        "DateTimeX::Easy" => "0.085",
        "Moose" => "0.41",
        "MooseX::Types" => "0.04",
        "MooseX::Types::DateTime" => "0.07",
        "MooseX::Types::Moose" => "0.04",
        "Time::Duration::Parse" => "0.06",
        "if" => 0,
        "namespace::clean" => "0.08",
        "perl" => "5.006",
        "strict" => 0,
        "warnings" => 0
      },
      "script_files" => [],
      "test_requires" => {
        "ExtUtils::MakeMaker" => 0,
        "File::Spec" => 0,
        "Test::Fatal" => 0,
        "Test::More" => 0,
        "ok" => 0
      }
    );


    my %fallback_build_requires = (
      "ExtUtils::MakeMaker" => 0,
      "File::Spec" => 0,
      "Test::Fatal" => 0,
      "Test::More" => 0,
      "ok" => 0
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);

    $build->create_build_script;
}


lib/MooseX/Types/DateTime/MoreCoercions.pm
use strict;
use warnings;
package MooseX::Types::DateTime::MoreCoercions;
BEGIN {
  $MooseX::Types::DateTime::MoreCoercions::AUTHORITY = 'cpan:ILMARI';
}
# git description: v0.13-2-g38c0d44
$MooseX::Types::DateTime::MoreCoercions::VERSION = '0.14';
# ABSTRACT: Extensions to L<MooseX::Types::DateTime>
# KEYWORDS: moose types constraints coercions date time datetime timestamp

use Moose 0.41 ();
use DateTime 0.4302 ();
use DateTime::Duration 0.4302 ();
use DateTimeX::Easy 0.085 ();
use Time::Duration::Parse 0.06 qw(parse_duration);
use MooseX::Types::DateTime 0.07 ();
use MooseX::Types::Moose 0.04 qw/Num HashRef Str/;
use namespace::clean 0.08;

use MooseX::Types 0.04 -declare => [qw( DateTime Duration)];
use if MooseX::Types->VERSION >= 0.42, 'namespace::autoclean';

subtype DateTime, as MooseX::Types::DateTime::DateTime;
coerce( DateTime,
    @{ MooseX::Types::DateTime::DateTime->coercion->type_coercion_map },
    from Str, via { DateTimeX::Easy->new($_) },
);

subtype Duration, as MooseX::Types::DateTime::Duration;
coerce( Duration,
    @{ MooseX::Types::DateTime::Duration->coercion->type_coercion_map },
    from Str, via {
        DateTime::Duration->new(
            seconds => parse_duration($_)
        );
    },
);

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

MooseX::Types::DateTime::MoreCoercions - Extensions to L<MooseX::Types::DateTime>

=head1 VERSION

version 0.14

=head1 SYNOPSIS

    package MyApp::MyClass;

    use MooseX::Types::DateTime::MoreCoercions qw( DateTime );

    has created => (
        isa => DateTime,
        is => "rw",
        coerce => 1,
    );

    my $instance = MyApp::MyClass->new(created=>'January 1, 1980');
    print $instance->created->year; # is 1980

    ## Coercions from the base type continue to work as normal.
    my $instance = MyApp::MyClass->new(created=>{year=>2000,month=>1,day=>10});

Please see the test case for more example usage.

=head1 DESCRIPTION

This module builds on L<MooseX::Types::DateTime> to add additional custom types and coercions.  Since it builds on an existing type, all coercions and constraints are inherited.

=head1 SUBTYPES

This module defines the following additional subtypes.

=head2 DateTime

Subtype of L<MooseX::Types::DateTime/DateTime>.  Adds an additional coercion from strings.

Uses L<DateTimeX::Easy> to try and convert strings, like "yesterday" into a valid L<DateTime> object.  Please note that due to ambiguity with how different systems might localize their timezone, string parsing may not always return the most expected value.  IN general we try to localize to UTC whenever possible.  Feedback welcomed!

=head2 Duration

Subtype of L<MooseX::Types::DateTime/Duration> that coerces from a string.  We use the module L<Time::Duration::Parse> to attempt this.

=head1 CAVEATS

Firstly, this module uses L<DateTimeX::Easy> which is way more DWIM than any sane person would desire. L<DateTimeX::Easy> works by falling back until something makes sense, this is variable. Furthermore, all the modules that L<DateTimeX::Easy> *can* use aren't required for "proper" function of L<DateTimeX::Easy>. What does this mean? Simple, your mileage may vary in your coercions because L<DateTimeX::Easy> is installation specific.

=head1 SEE ALSO

=over 4

=item * L<MooseX::Types::DateTime> Replacement for this module -- coercions with less voodoo

=item * L<DateTimeX::Easy> Backend of this module

=item * L<Time::Duration::Parse> Duration parsing backend for this module

=back

=head1 AUTHOR

John Napiorkowski E<lt>jjn1056 at yahoo.comE<gt>

Broken into a separate package from L<MooseX::Types::DateTime> by Evan Carroll.

Forked from L<MooseX::Types::DateTimeX> and ported back to use
L<MooseX::Types::DateTime> by Dagfinn Ilmari Mannsåker
E<lt>ilmari@ilmari.orgE<gt>.

=head1 AUTHORS

=over 4

=item *

John Napiorkowski <jjn1056@yahoo.com>

=item *

Dagfinn Ilmari Mannsåker <ilmari@ilmari.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2008 by John Napiorkowski.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 CONTRIBUTORS

=for stopwords Dagfinn Ilmari Mannsåker Evan Carroll Karen Etheridge gregor herrmann root

=over 4

=item *

Dagfinn Ilmari Mannsåker <ilmari@photobox.com>

=item *

Evan Carroll <me+cpan@evancarroll.com>

=item *

Evan Carroll <me@evancarroll.com>

=item *

Karen Etheridge <ether@cpan.org>

=item *

gregor herrmann <gregoa@debian.org>

=item *

root <root@x60s.(none)>

=back

=cut


Makefile.PL
# This Makefile.PL for MooseX-Types-DateTime-MoreCoercions was generated by
# Dist::Zilla::Plugin::MakeMaker::Fallback 0.010
# and Dist::Zilla::Plugin::MakeMaker::Awesome 0.23.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

BEGIN {
my %configure_requires = (
);

my @missing = grep {
    ! eval "require $_; $_->VERSION($configure_requires{$_}); 1"
} keys %configure_requires;

if (not @missing)
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";
}
else
{
    $ENV{PERL_MM_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

If you're installing manually, please retrain your fingers to run Build.PL
when present instead.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are doing something overly clever, in which case you
should consider setting the 'prefer_installer' config option in CPAN.pm, or
'prefer_makefile' in CPANPLUS, to 'mb" and '0' respectively.

You can also silence this warning for future installations by setting the
PERL_MM_FALLBACK_SILENCE_WARNING environment variable.

EOW

    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
}
}

use 5.006;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Extensions to L<MooseX::Types::DateTime>",
  "AUTHOR" => "John Napiorkowski <jjn1056\@yahoo.com>, Dagfinn Ilmari Manns\x{e5}ker <ilmari\@ilmari.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build::Tiny" => "0.037"
  },
  "DISTNAME" => "MooseX-Types-DateTime-MoreCoercions",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "MooseX::Types::DateTime::MoreCoercions",
  "PREREQ_PM" => {
    "DateTime" => "0.4302",
    "DateTime::Duration" => "0.4302",
    "DateTimeX::Easy" => "0.085",
    "Moose" => "0.41",
    "MooseX::Types" => "0.04",
    "MooseX::Types::DateTime" => "0.07",
    "MooseX::Types::Moose" => "0.04",
    "Time::Duration::Parse" => "0.06",
    "if" => 0,
    "namespace::clean" => "0.08",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::Fatal" => 0,
    "Test::More" => 0,
    "ok" => 0
  },
  "VERSION" => "0.14",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "DateTime" => "0.4302",
  "DateTime::Duration" => "0.4302",
  "DateTimeX::Easy" => "0.085",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Moose" => "0.41",
  "MooseX::Types" => "0.04",
  "MooseX::Types::DateTime" => "0.07",
  "MooseX::Types::Moose" => "0.04",
  "Test::Fatal" => 0,
  "Test::More" => 0,
  "Time::Duration::Parse" => "0.06",
  "if" => 0,
  "namespace::clean" => "0.08",
  "ok" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



